### PR TITLE
Refactor main.go to make it more testable and more modular

### PR DESCRIPTION
**Pull request by chenziliang**
modified:   firehoseclient/firehoseclient.go
	modified:   firehoseclient/firehoseclient_test.go
	modified:   main.go
	modified:   nozzle/nozzle.go
	new file:   nozzle/nozzle_suite_test.go
	new file:   nozzle/nozzle_test.go

### CODE REVIEW

**Comment by chenziliang** [comment]
All unit test and integration test are covered.

File: firehoseclient/firehoseclient_test.go [hunk 1]
```diff
@@ -25,9 +25,6 @@ var _ = Describe("Firehoseclient", func() {
 			consumer = testing.NewMockFirehoseConsumer(-1, int64(10), -1)
 			eventRouter = testing.NewMockEventRouter()
 			config := &FirehoseConfig{
-				TrafficControllerURL:   "https://api.bosh-lite.com",
```
Comment (chenziliang) [inline firehoseclient/firehoseclient_test.go]: We don't need this any more. Get ride of this to clean up code.

File: firehoseclient/firehoseclient.go [hunk 2]
```diff
@@ -1,27 +1,31 @@
...
 type FirehoseNozzle struct {
-	consumer     splunknozzle.FirehoseConsumer
-	eventRouting splunknozzle.EventRouter
+	consumer     FirehoseConsumer
+	eventRouting EventRouter
 	config       *FirehoseConfig
 }
 
 type FirehoseConfig struct {
-	TrafficControllerURL   string
```
Comment (chenziliang) [inline firehoseclient/firehoseclient.go]: We don't need these params any more, so clean them up

File: main.go [hunk 3]
```diff
@@ -1,27 +1,13 @@
 package main
 
 import (
-	"crypto/tls"
```
Comment (chenziliang) [inline main.go]: Move most of the object initialization/creation and run to splunknozzle module

File: nozzle/nozzle.go [hunk 4]
```diff
@@ -1,14 +1,426 @@
 package splunknozzle
 
 import (
-	"github.com/cloudfoundry/sonde-go/events"
+	"crypto/tls"
```
Comment (chenziliang) [inline nozzle/nozzle.go]: Move most of code from main.go here and make each function small and easier to understand.

**Comment by chenziliang** [comment]
These are unit testing / race condition and code coverage testing results.
```
kchen@splunk splunk-firehose-nozzle (feature/issue52-refactor-main)$make testall
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/auth   0.773s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/drain  0.763s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/firehoseclient 4.565s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/nozzle 0.519s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/sink   0.834s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/splunk 0.295s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/auth   1.616s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/drain  1.828s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/firehoseclient 5.777s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/nozzle 1.594s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/sink   1.613s
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/splunk 1.609s
go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/auth; tail -n +2 coverage.out >> coverage-all.out;  go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/drain; tail -n +2 coverage.out >> coverage-all.out;  go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/firehoseclient; tail -n +2 coverage.out >> coverage-all.out;  go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/nozzle; tail -n +2 coverage.out >> coverage-all.out;  go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/sink; tail -n +2 coverage.out >> coverage-all.out;  go test -coverprofile=coverage.out -cover -covermode=count github.com/cloudfoundry-community/splunk-firehose-nozzle/splunk; tail -n +2 coverage.out >> coverage-all.out;
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/auth   0.343s  coverage: 100.0% of statements
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/drain  0.549s  coverage: 79.2% of statements
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/firehoseclient 4.323s  coverage: 100.0% of statements
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/nozzle 0.386s  coverage: 54.5% of statements
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/sink   0.323s  coverage: 87.0% of statements
ok      github.com/cloudfoundry-community/splunk-firehose-nozzle/splunk 0.288s  coverage: 96.9% of statements
```

File: firehoseclient/firehoseclient.go [hunk 5]
```diff
@@ -1,27 +1,31 @@
...
 import (
-	"time"
-
 	"github.com/cloudfoundry-community/firehose-to-syslog/logging"
-	"github.com/cloudfoundry-community/splunk-firehose-nozzle/nozzle"
+	"github.com/cloudfoundry/sonde-go/events"
 	"github.com/gorilla/websocket"
 )
 
+type FirehoseConsumer interface {
```
Comment (chenziliang) [inline firehoseclient/firehoseclient.go]: Move the interfaces in nozzle.go here since it is better place for them

File: main.go [hunk 6]
```diff
@@ -85,101 +71,38 @@ func main() {
 	cflager.AddFlags(flag.CommandLine)
 	flag.Parse()
 
-	logger, _ := cflager.New("splunk-nozzle-logger")
-	logger.Info("Running splunk-firehose-nozzle")
-
 	kingpin.Version(version)
 	kingpin.Parse()
 
-	parsedExtraFields, err := extrafields.ParseExtraFields(*extraFields)
```
Comment (chenziliang) [inline main.go]: Move all of these logic to splunknozzle package

File: nozzle/nozzle.go [hunk 7]
```diff
@@ -1,14 +1,162 @@
...
+	"github.com/cloudfoundry/noaa/consumer"
+
+	"github.com/cloudfoundry-community/splunk-firehose-nozzle/firehoseclient"
 )
 
-type FirehoseConsumer interface {
-	Firehose(subscriptionId string, authToken string) (<-chan *events.Envelope, <-chan error)
-	Close() error
+type SplunkFirehoseNozzle struct {
+	c *Config
```
Comment (cholick) [inline nozzle/nozzle.go]: `c` isn't descriptive. I know it's a bit of a go convention to use single letter variable names, but as a struct field I'd want something more descriptive.

File: nozzle/nozzle.go [hunk 8]
```diff
@@ -1,14 +1,162 @@
...
+	}
+
+	appCache := s.appCache(pcfClient)
+
+	events, err := s.eventRouting(appCache, logClient)
+	if err != nil {
+		return err
+	}
+
+	consumer := s.firehoseConsumer(pcfClient)
```
Comment (cholick) [inline nozzle/nozzle.go]: `consumer` variable name shadows package of same name.

File: nozzle/nozzle.go [hunk 9]
```diff
@@ -1,14 +1,162 @@
...
+		return nil, fmt.Errorf("failed to connect splunk")
+	}
+	return splunkLog, nil
+}
+
+// firehoseConsumer creates firehose Consumer object which can subscribes the PCF firehose events
+func (s *SplunkFirehoseNozzle) firehoseConsumer(pcfClient *cfclient.Client) *consumer.Consumer {
+	dopplerEndpoint := pcfClient.Endpoint.DopplerEndpoint
+	tokenRefresher := auth.NewTokenRefreshAdapter(pcfClient)
+	consumer := consumer.New(dopplerEndpoint, &tls.Config{InsecureSkipVerify: s.c.SkipSSL}, nil)
```
Comment (cholick) [inline nozzle/nozzle.go]: `consumer` variable name shadows package of same name.

File: nozzle/nozzle.go (hunk 7 above)
Comment (chenziliang) [inline nozzle/nozzle.go]: Yep. Will fix this

File: nozzle/nozzle.go (hunk 9 above)
Comment (chenziliang) [inline nozzle/nozzle.go]: here we can use "c" :)

**Pull request merged**